05-dijkstras/06-dijkstras-vs-negative-graphs.md

```diff
@@ -0,0 +1,127 @@
+# Optional: Dijkstra vs. Negative Weighted Graphs
+
+### !callout
+
+## Optional Lesson
+
+This lesson is optional because it is not necessary to learn Dijkstra's Algorithm. However, it is does lend a handing in understanding the algorithm.
+
+### !end-callout
+
+As stated previously in the lesson, one of the limitations of Dijkstra’s Algorithm is that when implemented on a graph that contains negatively weighted edges it can possibly produce inaccurate results. This is because of the [“greedy”](https://en.wikipedia.org/wiki/Greedy_algorithm) nature of the algorithm, meaning that it works under the assumption that each subpath it chooses is the most optimal path. Therefore when we visit a node, Dijkstra’s algorithm assumes the current least costly path to the node locally will be the least costly path to the node globally.
+
+Once we’ve visited a node, it is considered closed and there is no way to re-open the path to that node and further relax the cost of the path. Dijkstra is designed to be applied to graphs with no negative costs. Such graphs can be used to solve many useful problems, so this limitation (the same we have with DFS and BFS) doesn't pose a problem when used as designed. If negative weights need to be handled, then the assumptions no longer hold, and some modification needs to be applied. 
+
+Let's walk through an example of Dijkstra’s Algorithm being implemented on a small negatively weighted graph to find the shortest path from Node A to Node D.
+
+![Dijkstra vs.Negative Graph-1](images/negative-1.png)
+
+Step 0. We start with a weighted graph. Notice that the weight on the edge connecting B to C is -100. The graph is directed. However, this is arbitrary.
+
+![Dijkstra vs.Negative Graph-2](images/negative-2.png)
+
+Step 1. We identify our start node, Node A, and set the cost to travel from the start node to all other nodes to infinity. The cost to travel to Node A (the start node itself) is set to 0. Something else that will prove to be useful is a list that keeps track of the nodes we’ve already visited. Finally, we add Node A to the top of our priority queue.
+
+```text
+priority queue:
+	- A
```

Comment:
Weird, the indentation looks off here still (and a little later on), but in the version I pulled down it looks OK?